File: java/05-reactor/src/test/java/com/fResult/reactor/ch5_10/OnErrorReturnTest.java
package com.fResult.reactor.ch5_10;

import lombok.extern.log4j.Log4j2;
import lombok.val;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Flux;
import reactor.test.StepVerifier;

@Log4j2
class OnErrorReturnTest {
  private final Flux<Integer> resultsInError = Flux.just(1, 2, 3).flatMap(this::failOn2);

  @Test
  void onErrorReturn() {
    val integers = resultsInError.onErrorReturn(0);
    StepVerifier.create(integers).expectNext(1, 0).verifyComplete();
  }

  private Flux<Integer> failOn2(int counter) {
    return counter == 2 ? Flux.error(new IllegalArgumentException("Oops!")) : Flux.just(counter);
  }
}

File: kotlin/05-reactor/src/test/kotlin/com/fResult/reactor/ch5_10/OnErrorReturnTest.kt
package com.fResult.reactor.ch5_10

import reactor.core.publisher.Flux
import reactor.test.StepVerifier
import kotlin.test.Test

class OnErrorReturnTest {
  private val resultsInError = Flux.just(1, 2, 3).map(::failOn2)

  @Test
  fun onErrorReturn() {
    val integers = resultsInError.onErrorReturn(RuntimeException::class.java, 0)

    StepVerifier.create(integers).expectNext(1, 0).verifyComplete()
  }

  private fun failOn2(counter: Int): Int = if (counter == 2)
    throw RuntimeException("Error on value $counter")
  else counter
}
